File: ToolGood.Algorithm/Base64.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

namespace ToolGood.Algorithm
{
    /// <summary>
    /// Modified Base64 for URL applications ('base64url' encoding)
    /// 
    /// See http://tools.ietf.org/html/rfc4648
    /// For more information see http://en.wikipedia.org/wiki/Base64
    /// </summary>
    static class Base64
    {

        public static string ToBase64String(byte[] input)
        {
            return Convert.ToBase64String(input);
        }

        public static byte[] FromBase64String(string base64)
        {
            return FromBase64ForUrlString(base64);
        }


        /// <summary>
        /// Modified Base64 for URL applications ('base64url' encoding)
        /// 
        /// See http://tools.ietf.org/html/rfc4648
        /// For more information see http://en.wikipedia.org/wiki/Base64
        /// </summary>
        /// <param name="input"></param>
        /// <returns>Input byte array converted to a base64ForUrl encoded string</returns>
        public static string ToBase64ForUrlString(byte[] input)
        {
            StringBuilder result = new StringBuilder(Convert.ToBase64String(input).TrimEnd('='));
            result.Replace('+', '-');
            result.Replace('/', '_');
            return result.ToString();
        }
        /// <summary>
        /// Modified Base64 for URL applications ('base64url' encoding)
        /// 
        /// See http://tools.ietf.org/html/rfc4648
        /// For more information see http://en.wikipedia.org/wiki/Base64
        /// </summary>
        /// <param name="base64ForUrlInput"></param>
        /// <returns>Input base64ForUrl encoded string as the original byte array</returns>
        private const string base64 = "===========================================+=+=/0123456789=======ABCDEFGHIJKLMNOPQRSTUVWXYZ====/=abcdefghijklmnopqrstuvwxyz=====";
        public static byte[] FromBase64ForUrlString(string base64ForUrlInput)
        {
            StringBuilder sb = new StringBuilder();
            foreach (var c in base64ForUrlInput) {
                if ((int)c >= 128) continue;
                var k = base64[c];
                if (k == '=') continue;
                sb.Append(k);
            }
            var len = sb.Length;
            int padChars = (len % 4) == 0 ? 0 : (4 - (len % 4));
            if (padChars > 0) sb.Append(String.Empty.PadRight(padChars, '='));
            return Convert.FromBase64String(sb.ToString());

            //base64ForUrlInput = Regex.Replace(base64ForUrlInput, @"[^a-zA-Z0-9+/\-_]", "");

            //int padChars = (base64ForUrlInput.Length % 4) == 0 ? 0 : (4 - (base64ForUrlInput.Length % 4));
            //StringBuilder result = new StringBuilder(base64ForUrlInput, base64ForUrlInput.Length + padChars);
            //result.Append(String.Empty.PadRight(padChars, '='));
            //result.Replace('-', '+');
            //result.Replace('_', '/');
            //return Convert.FromBase64String(result.ToString());
        }
    }
}

File: JavaScript/src/Internals/Functions/CsharpWeb/Base64.js
// Modified Base64 for URL applications ('base64url' encoding)
//
// See http://tools.ietf.org/html/rfc4648
// For more information see http://en.wikipedia.org/wiki/Base64

const base64 = "===========================================+=+=/0123456789=======ABCDEFGHIJKLMNOPQRSTUVWXYZ====/=abcdefghijklmnopqrstuvwxyz=====";

class Base64 {
    static ToBase64String(input) {
        return btoa(String.fromCharCode(...input));
    }

    static FromBase64String(base64Str) {
        return Base64.FromBase64ForUrlString(base64Str);
    }

    // Modified Base64 for URL applications ('base64url' encoding)
    //
    // See http://tools.ietf.org/html/rfc4648
    // For more information see http://en.wikipedia.org/wiki/Base64
    //
    // @param {Uint8Array} input
    // @returns {string} Input byte array converted to a base64ForUrl encoded string
    static ToBase64ForUrlString(input) {
        let result = btoa(String.fromCharCode(...input)).replace(/=/g, '');
        result = result.replace(/\+/g, '-');
        result = result.replace(/\//g, '_');
        return result;
    }

    // Modified Base64 for URL applications ('base64url' encoding)
    //
    // See http://tools.ietf.org/html/rfc4648
    // For more information see http://en.wikipedia.org/wiki/Base64
    //
    // @param {string} base64ForUrlInput
    // @returns {Uint8Array} Input base64ForUrl encoded string as the original byte array
    static FromBase64ForUrlString(base64ForUrlInput) {
        const sb = [];
        for (const c of base64ForUrlInput) {
            if (c.charCodeAt(0) >= 128) continue;
            const k = base64[c.charCodeAt(0)];
            if (k === '=') continue;
            sb.push(k);
        }
        const len = sb.length;
        const padChars = len % 4 === 0 ? 0 : 4 - (len % 4);
        if (padChars > 0) {
            sb.push(...'='.repeat(padChars).split(''));
        }
        const base64Str = sb.join('');
        const binaryString = atob(base64Str);
        const bytes = new Uint8Array(binaryString.length);
        for (let i = 0; i < binaryString.length; i++) {
            bytes[i] = binaryString.charCodeAt(i);
        }
        return bytes;
    }
}

export { Base64 };
